pkg/query-service/app/integrations/messagingQueues/kafka/sql.go

```diff
@@ -40,6 +40,257 @@ ORDER BY
+    COALESCE(total_requests / %d, 0) AS throughput
+FROM
+    consumer_latency
+`, start, end, service, queueType, topic, timeRange)
+	return query
+}
+
+// S3 consumer overview
+func generateConsumerLatencySQL(start, end int64, queueType string) string {
+	timeRange := (end - start) / 1000000000
+	query := fmt.Sprintf(`
+WITH consumer_latency AS (
+    SELECT
+        serviceName,
+        stringTagMap['messaging.destination.name'] AS topic,
+        quantile(0.99)(durationNano) / 1000000 AS p99,
+        COUNT(*) AS total_requests,
+        sumIf(1, statusCode = 2) AS error_count,
+        SUM(numberTagMap['messaging.message.body.size']) AS total_bytes
+    FROM signoz_traces.distributed_signoz_index_v2
+    WHERE
+        timestamp >= '%d'
+        AND timestamp <= '%d'
+        AND kind = 5
+        AND msgSystem = '%s'
+    GROUP BY topic, serviceName
+)
+
+SELECT
+    topic,
+    serviceName AS service_name,
+    p99,
+    COALESCE((error_count * 100.0) / total_requests, 0) AS error_rate,
+    COALESCE(total_requests / %d, 0) AS ingestion_rate,
+    COALESCE(total_bytes / %d, 0) AS byte_rate
+FROM
+    consumer_latency
+ORDER BY
+    topic;
+`, start, end, queueType, timeRange, timeRange)
+	return query
+}
+
+// S3 consumer topic/service
+func generateConsumerServiceLatencySQL(start, end int64, topic, service, queueType string) string {
+	timeRange := (end - start) / 1000000000
+	query := fmt.Sprintf(`
+WITH consumer_latency AS (
+    SELECT
+        quantile(0.99)(durationNano) / 1000000 AS p99,
+		stringTagMap['messaging.destination.partition.id'] AS partition,
+        COUNT(*) AS total_requests,
+        sumIf(1, statusCode = 2) AS error_count
+    FROM signoz_traces.distributed_signoz_index_v2
+    WHERE
+        timestamp >= '%d'
+        AND timestamp <= '%d'
+        AND kind = 5
+		AND serviceName = '%s'
+        AND msgSystem = '%s'
+		AND stringTagMap['messaging.destination.name'] = '%s'
+    GROUP BY partition
+)
+
+SELECT
+	partition,
+    p99,
+    COALESCE((error_count * 100.0) / total_requests, 0) AS error_rate,
+    COALESCE(total_requests / %d, 0) AS throughput
+FROM
+    consumer_latency
+`, start, end, service, queueType, topic, timeRange)
+	return query
+}
+
+// s4
+func generateProducerConsumerEvalSQL(start, end int64, queueType string, evalTime int64) string {
+	query := fmt.Sprintf(`
+WITH trace_data AS (
+    SELECT
+        p.serviceName AS producer_service,
+        c.serviceName AS consumer_service,
+        p.traceID,
+        p.timestamp AS producer_timestamp,
+        c.timestamp AS consumer_timestamp,
+        p.durationNano AS durationNano,
+        (toUnixTimestamp64Nano(c.timestamp) - toUnixTimestamp64Nano(p.timestamp)) + p.durationNano AS time_difference
+    FROM
+        signoz_traces.distributed_signoz_index_v2 p
+    INNER JOIN
+        signoz_traces.distributed_signoz_index_v2 c
+            ON p.traceID = c.traceID
+            AND c.parentSpanID = p.spanID
```

Comment:
are we sure that a consumer span must have a producer span as the parent? Eg how do we handle span links?

Reply:
let me address this in https://github.com/SigNoz/engineering-pod/issues/1916